### PR TITLE
core/graph: build DotGeneral gradients as gemms, not multiply+reduce

### PR DESCRIPTION
**Pull request by guygrigsby**
## Problem

`dotGeneralVJP` builds each operand's gradient by broadcasting the *other* operand, multiplying elementwise, and `ReduceSum`-ing over the contracted axes:

```go
thisVJP = Mul(thisVJP, otherProjected)
thisVJP = ReduceSum(thisVJP, ...otherCrossAxes)
```

That is mathematically a matmul, but it is handed to the backend as an elementwise multiply + reduction. XLA pattern-matches *some* of these back into gemms (the activation/input gradients), but for **weight gradients** — which reduce the large batch dimension `B*T` into a small `[in, out]` output — it leaves them as **CUDA-core reductions**, not tensor-core gemms.

So in a transformer the forward and the input gradients run on tensor cores, but **every weight gradient runs ~18x slower** on CUDA cores. Since a training step is dominated by weight gradients, the whole backward is pathologically slow while the forward is fast.

### Evidence (100M-param Llama, bf16, seq 2048, batch 2, RTX 3070 Ti / sm_86, XLA GPU)

Optimized HLO, weight-gradient-shaped outputs (`[768,2048]`, `[2048,768]`):

| | tensor-core `gemm_fusion` | CUDA-core `kInput` reduce |
|---|---|---|
| before | 0 | 60 |
| after | all | 0 |

The forward pass was already fast (~26 ms); the entire gap was the backward. Eager PyTorch is fast at the identical batch because cuBLAS computes `dW` as a gemm with a transpose flag.

| | step time | tok/s |
|---|---|---|
| before | 3.04 s | 1,346 |
| **after** | **0.16 s** | **25,217** |
| eager PyTorch, same GPU | 0.13 s | 32,093 |

**~18.7x faster**, ~80% of eager PyTorch.

## Fix

Emit a `DotGeneral` for the gradient when the "other" operand has cross (free) axes — the common matmul case — so the backend lowers it to a gemm. Keep the original broadcast-multiply path for the no-cross case (e.g. a vector dot product), which `DotGeneral` cannot express (removing it fails `TestGradientDot`). The result layout is unchanged, so the transpose-back is unchanged.

## Verification

- Full `core/graph` gradient / dot / einsum suite passes unchanged (numerics identical).
- Added `TestDotGeneralVJPLowersToGemm`: asserts a matmul gradient builds `DotGeneral` nodes (forward + dA + dW), guarding against a silent regression back to multiply+reduce.
- End-to-end: a 100M Llama overfit test converges to the identical loss; full training step is 18.7x faster.

---

*Disclosure: the diagnosis and implementation of this change were produced with heavy involvement of an AI coding assistant.*

### CODE REVIEW

**Comment by janpfeifer** [comment]
Omg, this is embarrassing. Thank you so much for the fix!!
